kong/dynamic_hook/wrap_function_gen.lua

```diff
@@ -0,0 +1,221 @@
+  
+        local enabled_groups = dynamic_hook.enabled_groups
+        if not enabled_groups[group_name] then
+          return original_func(%s)
+        end
+      end
+  
+      if handlers.before_mut then
+        local ok
+        ok, %s = pcall(handlers.before_mut, %s)
+        if not ok then
+          ngx.log(ngx.WARN,
+                  string.format("failed to run before_mut hook of %%s: %%s",
+                              group_name, a0))
+        end
+      end
+  
+      if handlers.befores then
+        for _, func in ipairs(handlers.befores) do
+          local ok, err = pcall(func, %s)
+          if not ok then
+            ngx.log(ngx.WARN,
+                    string.format("failed to run before hook of %%s: %%s",
+                                  group_name, err))
+          end
+        end
+      end
+      
+      local r0, r1, r2, r3, r4, r5, r6, r7 = original_func(%s)
+  
+      if handlers.after_mut then
+        local ok, err = pcall(handlers.after_mut, r0, r1, r2, r3, r4, r5, r6, r7)
+        if not ok then
+          ngx.log(ngx.WARN,
+                  string.format("failed to run after_mut hook of %%s: %%s",
+                                group_name, err))
+        end
+      end
+  
+      if handlers.afters then
+        for _, func in ipairs(handlers.afters) do
+          local ok, err = pcall(func, r0, r1, r2, r3, r4, r5, r6, r7)
+          if not ok then
+            ngx.log(ngx.WARN,
+                    string.format("failed to run after hook of %%s: %%s",
+                                  group_name, err))
+          end
+        end
+      end
+  
+      return r0, r1, r2, r3, r4, r5, r6, r7
+    end
+  end
+]]
+
+
+local _M = {}
+
+
+local function warp_function_0(always_enabled_groups, group_name, original_func, handlers)
+  return function()
+    if not always_enabled_groups[group_name] then
+      local phase = ngx_get_phase()
+      if phase == "init" or phase == "init_worker" then
+        return original_func()
+      end
+
+      local dynamic_hook = ngx.ctx.dynamic_hook
+      if not dynamic_hook then
+        return original_func()
+      end
+
+      local enabled_groups = dynamic_hook.enabled_groups
+      if not enabled_groups[group_name] then
+        return original_func()
+      end
+    end
+
+    if handlers.before_mut then
+      local ok, err = pcall(handlers.before_mut)
+      if not ok then
+        ngx.log(ngx.WARN,
+                string.format("failed to run before_mut hook of %s: %s",
+                              group_name, err))
+      end
+    end
+
+    if handlers.befores then
+      for _, func in ipairs(handlers.befores) do
+        local ok, err = pcall(func)
+        if not ok then
+          ngx.log(ngx.WARN,
+                  string.format("failed to run before hook of %s: %s",
+                                group_name, err))
+        end
+      end
+    end
+
+    local r0, r1, r2, r3, r4, r5, r6, r7 = original_func()
+
+    if handlers.after_mut then
```

Comment:
I only noticed a reference to `.after_mut`, but I could'nt find  where it was assigned? Do I miss something?

Reply:
Currently, it is not used.